src/libraries/Common/src/Interop/Windows/Advapi32/Interop.TRUSTED_DOMAIN_FULL_INFORMATION.cs

```diff
@@ -0,0 +1,18 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Runtime.InteropServices;
+
+internal static partial class Interop
+{
+    internal static partial class Advapi32
+    {
+        [StructLayout(LayoutKind.Sequential)]
+        internal sealed class TRUSTED_DOMAIN_FULL_INFORMATION
```

Comment:
```suggestion
        internal struct TRUSTED_DOMAIN_FULL_INFORMATION
```

Reply:
Done in bf9983a0650062e3e51c81d9487c57ee892ae874

Before I resolve this, though, is this also now blittable and so should I change the `Marshal.SizeOf` calls to `sizeof` e.g. at https://github.com/chrisdcmoore/runtime/blob/285bbfa58772b9dfaff8ae3c6aae9a4ec1bc5dc6/src/libraries/System.DirectoryServices/src/System/DirectoryServices/ActiveDirectory/TrustHelper.cs#L661

Reply:
This has nullable types so I don't think so. If they were marked as pointers then it would be okay. You can try switching to `sizeof()`, but my guess is it will fail.

Reply:
I'm actually not sure why that field is nullable given its definition on the unmanaged side: https://learn.microsoft.com/en-us/windows/win32/api/ntsecapi/ns-ntsecapi-trusted_domain_full_information

Am I missing something or should it be not nullable?

I'm also not entirely clear on the differing visibility modifiers on its members either...

Reply:
(I'm still learning when it comes to interop code so I'm sorry if these are stupid questions!)

Reply:
> I'm sorry if these are stupid questions!

Definitely not. These questions are helpful and appreciated.

> I'm actually not sure why that field is nullable given its definition on the unmanaged side:

Hmmm. Agreed. This signature looks wrong. These don't seem to be nullable types. It looks like the type is incorrectly defined. I don't recall all the rules for `class`s marked as "Sequential" though so perhaps it is taking advantage of that. Given the documentation I think you should update it.

> I'm also not entirely clear on the differing visibility modifiers on its members either...

That is merely for C# and has nothing to do with the marshalling logic in this case.

/cc @jkoritzinsky @jtschuster

Reply:
Updated in ef237828a51d7efc0d09ea42dda53f1b8579da26 and now that they blittable, used sizeof in 142867362d4126f6c3346ae41896d0d82636f3c2.

Whilst doing some manual testing I've come across some pre-existing bugs. If I run the following code on a Windows machine joined to a domain ("MYLAB.LOCAL") which has a child domain ("SUB.MYLAB.LOCAL") then against `net48` the code works fine, but against `net8.0-windows` it throws an `ArgumentException`:

```c#
var domain = System.DirectoryServices.ActiveDirectory.Domain.GetCurrentDomain();
var isSelectiveAuthentication = domain.GetSelectiveAuthenticationStatus("sub.mylab.local");
```

I've tracked down the reason for this as calls to `Marshal.PtrToStructure(IntPtr, object)` where `object` is a struct rather than an instance of a class (`TRUSTED_DOMAIN_INFORMATION_EX`) - in full Framework code this is still a class, but has been turned into a struct at some point in the non-Framework code.

There are a few places in `TrustHelper.cs` where this mistake is present, for example:
https://github.com/dotnet/runtime/blob/654d5cd2ceb0693b3888e0a5a490a8f1a2ba356a/src/libraries/System.DirectoryServices/src/System/DirectoryServices/ActiveDirectory/TrustHelper.cs#L103

Would you like me to fix these bugs in this PR or open a separate issue to track (and PR to fix)? I'm conscious that the original intent of this PR was to lift and shift code into Common Interop only, with a view to making actual improvements, and fixes to the existing failing tests, in separate PRs to this so that they are each easy to review, but that is starting to no longer be the case.

Reply:
> Would you like me to fix these bugs in this PR or open a separate issue to track (and PR to fix)?

Yes please. These changes should reflect updating usage as appropriate.

> I've tracked down the reason for this as calls to Marshal.PtrToStructure(IntPtr, object) where object is a struct rather than an instance of a class (TRUSTED_DOMAIN_INFORMATION_EX)

Since the `TRUSTED_DOMAIN_INFORMATION_EX` is now a `struct`, the conversion can be as simple as converting/casting a pointer to the desired type.

Reply:
To be absolutely clear, that is a struct on `main`, i.e. before I touched any of this code - do you still want that bug fixed in this PR (in the more optimal way that you've suggested vs what I suggested?)

Reply:
I suggest to treat this as advanced topic and leave for next pr.
I plan to replace the `Marshal` related interop in next pr.

Reply:
I've elected to update the conversion code for structs that I've touched in 62d261f632610f188c8bbd15405648401f8ae3ad and to fix the pre-existing bugs in 1e6048b212a9150e878b2d88c2fac053e2c17135 - the latter did involve marking the containing methods (should I have used an unsafe block instead?).

It sounds like you have plans to overhaul a lot of this anyway @huoyaoyuan but I thought it best to change it here then at least the bug will be fixed until such a time.